components/autocomplete/index.js

```diff
@@ -176,40 +351,76 @@ class Autocomplete extends Component {
 		// Any handled keycode should prevent original behavior. This relies on
 		// the early return in the default case.
 		event.preventDefault();
-		event.stopImmediatePropagation();
+		event.stopPropagation();
+	}
+
+	toggleKeyEvents( isListening ) {
+		// This exists because we must capture ENTER key presses before Editable.
+		// It seems that react fires the simulated capturing events after the
+		// native browser event has already bubbled so we can't stopPropagation
+		//  and avoid Editable getting the event from TinyMCE, hence we must
+		// register a native event handler.
+		if ( ! this.node ) {
+			return;
+		}
+		const realNode = findDOMNode( this.node );
+		const handler = isListening ? 'addEventListener' : 'removeEventListener';
+		realNode[ handler ]( 'keydown', this.setSelectedIndex, true );
+	}
+
+	componentWillMount() {
+		const { completers } = this.props;
+		// Some completers must do asynchronous requests to get their options so
+		// we calculate that here.
+		completers.forEach( ( { getOptions }, idx ) => {
+			getOptions().then( ( options ) => {
+				this.allOptions[ idx ] = options.map( ( option, key ) => ( { ...option, key } ) );
+				this.forceUpdate();
+			} );
+		} );
+	}
+
+	componentDidUpdate( prevProps, prevState ) {
+		const { open } = this.state;
+		const { open: prevOpen } = prevState;
+		if ( ( open !== null ) !== ( prevOpen !== null ) ) {
+			this.toggleKeyEvents( open !== null );
+		}
+	}
+
+	componentWillUnmount() {
+		this.toggleKeyEvents( false );
 	}
 
 	render() {
-		const { children, className } = this.props;
-		const { isOpen, selectedIndex } = this.state;
+		const { children } = this.props;
+		const { open, selectedIndex, range } = this.state;
+		const { className } = open || {};
 		const classes = classnames( 'components-autocomplete__popover', className );
 		const filteredOptions = this.getFilteredOptions();
 
-		// Blur is applied to the wrapper node, since if the child is Editable,
-		// the event will not have `relatedTarget` assigned.
 		return (
 			<div
 				ref={ this.bindNode }
-				onBlur={ this.onBlur }
+				onInput={ this.search }
 				className="components-autocomplete"
 			>
-				{ cloneElement( Children.only( children ), {
-					onInput: this.search,
-					onKeyDown: this.setSelectedIndex,
-				} ) }
+				{ children }
 				<Popover
-					isOpen={ isOpen && filteredOptions.length > 0 }
+					isOpen={ filteredOptions.length > 0 }
 					focusOnOpen={ false }
+					onClickOutside={ () => this.reset() }
```

Comment:
Should this be binding to `onClose` ? We can't assume clicking outside is the only reason for a popover to close (escape press).

Reply:
I hadn't noticed `onClose` - I will investigate.

Reply:
It seems that onClose is no different to onClickOutside with exception of when the Popover is handling the ESCAPE key (which it is not in this case).